### PR TITLE
refactor and improve performance of table functions

### PR DESCRIPTION
**Pull request by BigSmoKe07**
Function Refactor: 
- contains
- table_matches
- table_merge

### CODE REVIEW

File: imports/table/shared.lua [hunk 1]
```diff
@@ -9,47 +9,54 @@ local pairs = pairs
...
 ---@param t2 any
 ---@return boolean
 ---Compares if two values are equal, iterating over tables and matching both keys and values.
 local function table_matches(t1, t2)
-	local type1, type2 = type(t1), type(t2)
-
-	if type1 ~= type2 then return false end
-	if type1 ~= 'table' and type2 ~= 'table' then return t1 == t2 end
+    if type(t1) ~= type(t2) then return false end
+    if type(t1) ~= 'table' then return t1 == t2 end
```
Comment (thelindat) [inline imports/table/shared.lua:44]: Considering we're micro-optimising here, there's no reason not to reuse type1.

Comment (thelindat) [reply imports/table/shared.lua:44]: While we're at it we can get a small boost by comparing table types and lengths.

```lua
    local tabletype1 = table.type(t1)

    if tabletype1 ~= table.type(t2) or (tabletype1 == 'array' and #t1 ~= #t2) then
        return false
    end
```

File: imports/table/shared.lua [hunk 2]
```diff
@@ -9,47 +9,57 @@ local pairs = pairs
...
 ---Compares if two values are equal, iterating over tables and matching both keys and values.
 local function table_matches(t1, t2)
-	local type1, type2 = type(t1), type(t2)
+    local tabletype1 = table.type(t1) 
 
-	if type1 ~= type2 then return false end
-	if type1 ~= 'table' and type2 ~= 'table' then return t1 == t2 end
+    if tabletype1 ~= table.type(t2) or (tabletype1 == 'array' and #t1 ~= #t2) then 
+        return false 
+    end
```
Comment (thelindat) [inline imports/table/shared.lua:47]: To clarify, I meant to use the table.type check in addition to the type check.
This _should_ work as a replacement for the previous code though.

```diff
- local type1 = type(t1)

- if type1 ~= type(t2) then return false end
- if type1 ~= 'table' then return t1 == t2 end

+ local tabletype1 = table.type(t1)

+ if not tabletype1 then return t1 == t2 end

+ if tabletype1 ~= table.type(t2) or (tabletype1 == 'array' and #t1 ~= #t2) then
+     return false
+ end
```

**Comment by rhykosqc** [comment]
With this update, we have encountered an inventory problem. We are no longer able to reload weapons.

--> I think its come from table.contains

```
function Inventory.GetSlotWithItem(itemName, metadata, strict)
    local inventory = PlayerData.inventory
    local item = Items(itemName) --[[@as OxClientItem?]]
    
    if not inventory or not item then return end
    
    metadata = assertMetadata(metadata)
    local tablematch = strict and table.matches or table.contains
    
    for _, slotData in pairs(inventory) do
        if slotData and slotData.name == item.name and (not metadata or tablematch(slotData.metadata, metadata)) then
	   return slotData
        end
    end
end
```

**Pull request merged**